connect/connect.go

```diff
@@ -0,0 +1,241 @@
+
+	return encodeComponents(servicesJSON, publicKey, redirectURL), nil
+}
+
+func GenerateQRCode(publicKey string, redirectURL string, input Services) (string, error) {
+	if publicKey == "" || redirectURL == "" || input == nil {
+		return "", &GandalfError{
+			Message: "Invalid input parameters",
+			Code:    QRCodeGenNotSupported,
+		}
+	}
+
+	services, err := runValidation(publicKey, redirectURL, input)
+	if err != nil {
+		return "", err
+	}
+
+	servicesJSON := servicesToJSON(services)
+	appClipURL := encodeComponents(servicesJSON, redirectURL, publicKey)
+
+	qrCode, err := qrcode.New(appClipURL, qrcode.Medium)
+	if err != nil {
+		return "", &GandalfError{
+			Message: "QRCode Generation Error",
+			Code:    QRCodeNotGenerated,
+		}
+	}
+
+	qrCodeData, err := qrCode.PNG(256)
+	if err != nil {
+		return "", &GandalfError{
+			Message: "QRCode Generation Error",
+			Code:    QRCodeNotGenerated,
+		}
+	}
+	qrCodeURL := fmt.Sprintf("data:image/png;base64,%s", base64.StdEncoding.EncodeToString(qrCodeData))
+	return qrCodeURL, nil
+}
+
+func introspectSauron() IntrospectionResult {
+	client := graphqlClient.NewClient(SAURON_BASE_URL)
+	req := graphqlClient.NewRequest(introspectionQuery)
+
+	ctx := context.Background()
+
+	var respData IntrospectionResult
+
+	if err := client.Run(ctx, req, &respData); err != nil {
+		log.Fatalf("Error making introspection query: %v", err)
+	}
+	return respData
+}
+
+func validateRedirectURL(rawURL string) error {
+	_, err := url.ParseRequestURI(rawURL)
+	if err != nil {
+		return &GandalfError{
+			Message: "Invalid redirect URL",
+			Code:    InvalidRedirectURL,
+		}
+	}
+	return nil
+}
+
+func validatePublicKey(publicKey string) bool {
+	return publicKeyRequest(publicKey)
+}
+
+func getSupportedServices() []Value {
+	gqlSchema := introspectSauron()
+	for _, val := range gqlSchema.Schema.Types {
+		if val.Kind == "ENUM" && val.Name == "Source" {
+			return val.EnumValues
+		}
+	}
+	return nil
+}
+
+func validateInputServices(input Services) (map[string]Service, error) {
+	// supportedServicesInterface := getSupportedServices()
+	supportedServices := getSupportedServices()
+	
+	serviceMap := make(map[string]Value)
+	for _, val := range supportedServices {
+		serviceMap[val.Name] = val
+	}
+
+	var unsupportedServices []string
+	requiredServices := 0
+	cleanedServices :=  make(map[string]Service)
+	for _, val := range input {
+		key := strings.ToUpper(val.Name)
+		if _, found := serviceMap[key]; !found {
+			unsupportedServices = append(unsupportedServices, key)
+			continue
+		}
+		
+		if val.Name != "" {
+			requiredServices++
+		}
+		cleanedServices[key] = val
+	}
+
+	if len(unsupportedServices) > 0 {
+		return nil, &GandalfError{
+			Message: fmt.Sprintf("These services %s are unsupported", strings.Join(unsupportedServices, " ")),
+			Code:    InvalidService,
+		}
+	}
+
+	if requiredServices < 1 {
+		return nil, &GandalfError{
+			Message: "At least one service has to be required",
```

Comment:
maybe? "At least one service is required" ?

Reply:
Yes at least a service has to be passed